bpf/lib/lb.h

```diff
@@ -847,24 +866,35 @@ lb6_select_backend_id(struct __ctx_buff *ctx __maybe_unused,
 # error "Invalid load balancer backend selection algorithm!"
 #endif /* LB_SELECTION */
 
-static __always_inline int lb6_xlate(struct __ctx_buff *ctx, __u8 nexthdr,
+static __always_inline int lb6_xlate(struct __ctx_buff *ctx,
+				     const union v6addr *new_saddr __maybe_unused,
+				     const union v6addr *old_saddr __maybe_unused,
+				     __u8 nexthdr,
 				     int l3_off, int l4_off,
 				     const struct lb6_key *key,
 				     const struct lb6_backend *backend,
 				     bool has_l4_header)
 {
 	const union v6addr *new_dst = &backend->address;
 	struct csum_offset csum_off = {};
+	__be32 sum;
 
 	if (has_l4_header)
 		csum_l4_offset_and_flags(nexthdr, &csum_off);
 
 	if (ipv6_store_daddr(ctx, new_dst->addr, l3_off) < 0)
 		return DROP_WRITE_ERROR;
-	if (csum_off.offset) {
-		__be32 sum = csum_diff(key->address.addr, 16, new_dst->addr,
-				       16, 0);
+	sum = csum_diff(key->address.addr, 16, new_dst->addr, 16, 0);
+#ifdef USE_LOOPBACK_LB
+	if (new_saddr && (new_saddr->d1 || new_saddr->d2)) {
+		cilium_dbg_lb(ctx, DBG_LB6_LOOPBACK_SNAT, old_saddr->p4, new_saddr->p4);
 
+		if (ipv6_store_saddr(ctx, (void *)new_saddr->addr, l3_off) < 0)
+			return DROP_WRITE_ERROR;
+		sum = csum_diff(old_saddr->addr, 16, new_saddr->addr, 16, 0);
```

Comment:
I believe we need to provide the previous `sum` to have a correct computation of the final checksum
```suggestion
		sum = csum_diff(old_saddr->addr, 16, new_saddr->addr, 16, sum);
```

Reply:
A slightly more performant version could be
```c
{
	const union v6addr *new_dst = &backend->address;
	struct csum_offset csum_off = {};
	__be32 sum = 0;

	if (has_l4_header)
		csum_l4_offset_and_flags(nexthdr, &csum_off);

	if (ipv6_store_daddr(ctx, new_dst->addr, l3_off) < 0)
		return DROP_WRITE_ERROR;
#ifdef USE_LOOPBACK_LB
	if (new_saddr && (new_saddr->d1 || new_saddr->d2)) {
		cilium_dbg_lb(ctx, DBG_LB6_LOOPBACK_SNAT, old_saddr->p4, new_saddr->p4);

		if (ipv6_store_saddr(ctx, (void *)new_saddr->addr, l3_off) < 0)
			return DROP_WRITE_ERROR;
		sum = csum_diff(old_saddr->addr, 16, new_saddr->addr, 16, 0);
	}
#endif /* USE_LOOPBACK_LB */
	if (csum_off.offset) {
		sum = csum_diff(key->address.addr, 16, new_dst->addr, 16, sum);

		if (csum_l4_replace(ctx, l4_off, &csum_off, 0, sum,
				    BPF_F_PSEUDO_HDR) < 0)
			return DROP_CSUM_L4;
	}

	if (!has_l4_header)
		return CTX_ACT_OK;

	return lb_l4_xlate(ctx, nexthdr, l4_off, &csum_off, key->dport,
			   backend->port);
}
```

in this way we compute `sum = csum_diff(key->address.addr, 16, new_dst->addr, 16, sum);` only when we really need to rewrite the l4 checksum. WDYT?

Reply:
I think your suggestion looks better. But, we may need to be consistent with `lb4_xlate`. I'd love to get other reviewers' opinions on this.

Reply:
> But, we may need to be consistent with lb4_xlate

IMO we should keep things consistent with the v4 stuff until this is merged. Then, if there's an optimization to be made, it could be done in another PR that targets both v4 and v6 at the same time.